volume/fs_test.go

```diff
@@ -0,0 +1,135 @@
+package volume
+
+import (
+	"context"
+	"testing"
+
+	"github.com/puppetlabs/wash/datastore"
+	"github.com/puppetlabs/wash/plugin"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/suite"
+)
+
+const varLogFixture = `
+96 1550611510 1550611448 1550611448 41ed /var/log/path
+96 1550611510 1550611448 1550611448 41ed /var/log/path/has
+96 1550611510 1550611448 1550611448 41ed /var/log/path/has/got
+96 1550611510 1550611458 1550611458 41ed /var/log/path/has/got/some
+0 1550611458 1550611458 1550611458 81a4 /var/log/path/has/got/some/legs
+96 1550611510 1550611453 1550611453 41ed /var/log/path1
+0 1550611453 1550611453 1550611453 81a4 /var/log/path1/a file
+96 1550611510 1550611441 1550611441 41ed /var/log/path2
+64 1550611510 1550611441 1550611441 41ed /var/log/path2/dir
+`
+
+type fsTestSuite struct {
+	suite.Suite
+}
+
+func (suite *fsTestSuite) SetupSuite() {
+	plugin.SetTestCache(datastore.NewMemCache())
+}
+
+func (suite *fsTestSuite) TearDownSuite() {
+	plugin.UnsetTestCache()
+}
+
+func createResult(data string) plugin.ExecResult {
+	outputch := make(chan plugin.ExecOutputChunk, 1)
+	execResult := plugin.ExecResult{
+		OutputCh:   outputch,
+		ExitCodeCB: func() (int, error) { return 0, nil },
+	}
+	outputch <- plugin.ExecOutputChunk{Data: data}
+	close(outputch)
+	return execResult
+}
+
+func createExec() *mockExecutor {
+	exec := &mockExecutor{EntryBase: plugin.NewEntry("instance")}
+	// Used when recording activity.
+	exec.SetTestID("/instance")
+	cmd := StatCmd("/var/log")
+	exec.On("Exec", mock.Anything, cmd[0], cmd[1:], plugin.ExecOptions{}).Return(createResult(varLogFixture), nil)
+	return exec
+}
+
+func (suite *fsTestSuite) dig(grp plugin.Group, names ...string) plugin.Entry {
+	entry := plugin.Entry(grp)
+	for _, name := range names {
+		entries, err := entry.(plugin.Group).List(context.Background())
+		suite.NoError(err)
+		for _, match := range entries {
+			if plugin.Name(match) == name {
+				entry = match
+				break
+			}
+		}
+	}
+	suite.Equal(names[len(names)-1], plugin.Name(entry))
+	return entry
+}
+
+func (suite *fsTestSuite) TestFSList() {
+	exec := createExec()
+	fs := NewFS("fs", exec)
+	// ID would normally be set when listing FS within the parent instance.
+	fs.SetTestID("/instance/fs")
+
+	entry := suite.dig(fs, "var", "log").(plugin.Group)
+	entries, err := entry.List(context.Background())
+	suite.NoError(err)
+	suite.Equal(3, len(entries))
+
+	suite.Equal("path", plugin.Name(entries[0]))
+	suite.Equal("path1", plugin.Name(entries[1]))
+	suite.Equal("path2", plugin.Name(entries[2]))
+	for _, entry := range entries {
+		_, ok := entry.(plugin.Group)
+		suite.True(ok)
```

Comment:
Think we'd want to fail the test if this assertion fails so that Lines 92 + 99 don't panic. May be worth doing a similar assertion in Line 79.